target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c

```diff
@@ -1335,7 +1335,7 @@ static int rtl838x_pie_rule_write(struct rtl838x_switch_priv *priv, int idx, str
 	/* Access IACL table (1) via register 0 */
 	struct table_reg *q = rtl_table_get(RTL8380_TBL_0, 1);
 	u32 r[18];
-	int err = 0;
+	int err;
```

Comment:
```suggestion
	int err = -EINVAL;
```

best to keep some initialization here. Anything is better then nothing :)

Reply:
@svanheule had suggested the exact opposite if I understood him correctly:
https://github.com/openwrt/openwrt/pull/11685#issuecomment-1370761187

Reply:
Looking at the entire function, we see that we guarantee assignment via https://github.com/openwrt/openwrt/blob/6783a8b8d872ca4507db987ed2e5e461fad9f291/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c#L1358 so then you can leave it uninitialized. Sorry for the noise :)

But always double check you don't end up with weird cases, you can return an uninitialized variable, that's why I personally prefer to always initialize things with known values (unless the first few lines immediately overwrites it).

Reply:
No worries, I had asked myself the same question, but had assume that maybe modern compilers can throw a warning if a function returns an uninitialized variable (but I'm absolutetly no expert for C compilers).

Reply:
It can, and it does. But have you actually looked at the output of our kernel build? I'm sure we miss a few warnings there :D

And I only looked at the code snippet :p so I only saw your variable declaration change, + the first use, which was inside an if-guard :)

Reply:
Thanks for the idea, this has actually made me try building with `CONFIG_WERROR=y` added to the kernel config, and fixing the stuff that's producing compiler warnings… 😄

Reply:
@hardfalcon Actually, I fixed almost all, if not all, of those warnings in my 170 large patch set. So 'in time' they should all drip in. I fixed them as part of the changes mostly, to not introduce too much ruckus :)

Reply:
These are the additional changes I had to do to make it compile with `CONFIG_WERROR=y` about an hour ago. Most of it is simply adding a `__attribute__((unused))` here and there until the respective functions and variables are actually used, but there were also 2-3 minor code changes apart from that.

```diff
diff --git a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/common.c b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/common.c
index 1fa92ae2..5f27d3a8 100644
--- a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/common.c
+++ b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/common.c
@@ -528,7 +528,7 @@ int rtl83xx_lag_del(struct dsa_switch *ds, int group, int port)
 }
 
 /* Allocate a 64 bit octet counter located in the LOG HW table */
-static int rtl83xx_octet_cntr_alloc(struct rtl838x_switch_priv *priv)
+static __attribute__((unused)) int rtl83xx_octet_cntr_alloc(struct rtl838x_switch_priv *priv)
 {
 	int idx;
 
@@ -1426,7 +1426,7 @@ static int rtl83xx_fib_event(struct notifier_block *this, unsigned long event, v
 			fib_info_hold(fib_work->fen_info.fi);
 
 		} else if (info->family == AF_INET6) {
-			struct fib6_entry_notifier_info *fen6_info = ptr;
+			struct fib6_entry_notifier_info __attribute__((unused)) *fen6_info = ptr;
 			pr_warn("%s: FIB_RULE ADD/DEL for IPv6 not supported\n", __func__);
 			kfree(fib_work);
 			return NOTIFY_DONE;
diff --git a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
index 83a0441c..db147e0e 100644
--- a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
+++ b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
@@ -2013,7 +2013,7 @@ static bool rtl83xx_lag_can_offload(struct dsa_switch *ds,
 
 static int rtl83xx_port_lag_change(struct dsa_switch *ds, int port)
 {
-	struct rtl838x_switch_priv *priv = ds->priv;
+	struct rtl838x_switch_priv __attribute__((unused)) *priv = ds->priv;
 
 	pr_debug("%s: %d\n", __func__, port);
 	/* Nothing to be done... */
diff --git a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c
index 504b2982..35a06d14 100644
--- a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c
+++ b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c
@@ -1291,7 +1291,7 @@ static void rtl838x_pie_rule_dump_raw(u32 r[])
 	pr_info("Sel    : %08x\n", r[17]);
 }
 
-static void rtl838x_pie_rule_dump(struct  pie_rule *pr)
+static __attribute__((unused)) void rtl838x_pie_rule_dump(struct  pie_rule *pr)
 {
 	pr_info("Drop: %d, fwd: %d, ovid: %d, ivid: %d, flt: %d, log: %d, rmk: %d, meter: %d tagst: %d, mir: %d, nopri: %d, cpupri: %d, otpid: %d, itpid: %d, shape: %d\n",
 		pr->drop, pr->fwd_sel, pr->ovid_sel, pr->ivid_sel, pr->flt_sel, pr->log_sel, pr->rmk_sel, pr->log_sel, pr->tagst_sel, pr->mir_sel, pr->nopri_sel,
diff --git a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c
index 7e4f13fb..b3fe5dbb 100644
--- a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c
+++ b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c
@@ -990,7 +990,7 @@ static u32 rtl930x_l3_hash4(u32 ip, int algorithm, bool move_dip)
 	return hash;
 }
 
-static u32 rtl930x_l3_hash6(struct in6_addr *ip6, int algorithm, bool move_dip)
+static __attribute__((unused)) u32 rtl930x_l3_hash6(struct in6_addr *ip6, int algorithm, bool move_dip)
 {
 	u32 rows[16];
 	u32 hash;
@@ -1409,7 +1409,7 @@ static void rtl930x_get_l3_nexthop(int idx, u16 *dmac_id, u16 *interface)
 	*interface = v & 0x7f;
 }
 
-static int rtl930x_l3_mtu_del(struct rtl838x_switch_priv *priv, int mtu)
+static __attribute__((unused)) int rtl930x_l3_mtu_del(struct rtl838x_switch_priv *priv, int mtu)
 {
 	int i;
 
@@ -1458,7 +1458,7 @@ static int rtl930x_l3_mtu_add(struct rtl838x_switch_priv *priv, int mtu)
 }
 
 /* Creates an interface for a route by setting up the HW tables in the SoC */
-static int rtl930x_l3_intf_add(struct rtl838x_switch_priv *priv, struct rtl838x_l3_intf *intf)
+static __attribute__((unused)) int rtl930x_l3_intf_add(struct rtl838x_switch_priv *priv, struct rtl838x_l3_intf *intf)
 {
 	int i, intf_id, mtu_id;
 	/* number of MTU-values < 16384 */
@@ -1681,7 +1681,7 @@ static void rtl930x_write_pie_templated(u32 r[], struct pie_rule *pr, enum templ
 	}
 }
 
-static void rtl930x_read_pie_fixed_fields(u32 r[], struct pie_rule *pr)
+static __attribute__((unused)) void rtl930x_read_pie_fixed_fields(u32 r[], struct pie_rule *pr)
 {
 	pr->stacking_port = r[6] & BIT(31);
 	pr->spn = (r[6] >> 24) & 0x7f;
diff --git a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl931x.c b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl931x.c
index 92f06622..f67545ad 100644
--- a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl931x.c
+++ b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl931x.c
@@ -1113,7 +1113,7 @@ static void rtl931x_write_pie_templated(u32 r[], struct pie_rule *pr, enum templ
 	}
 }
 
-static void rtl931x_read_pie_fixed_fields(u32 r[], struct pie_rule *pr)
+static __attribute__((unused)) void rtl931x_read_pie_fixed_fields(u32 r[], struct pie_rule *pr)
 {
 	pr->mgnt_vlan = r[7] & BIT(31);
 	if (pr->phase == PHASE_IACL)
diff --git a/target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c b/target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c
index 857d389f..c0447ac4 100644
--- a/target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c
+++ b/target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c
@@ -161,7 +161,7 @@ static void rtl931x_create_tx_header(struct p_hdr *h, unsigned int dest_port, in
 		h->cpu_tag[2] = (BIT(5) | (prio & 0x1f)) << 8;
 }
 
-static void rtl93xx_header_vlan_set(struct p_hdr *h, int vlan)
+static __attribute__((unused)) void rtl93xx_header_vlan_set(struct p_hdr *h, int vlan)
 {
 	h->cpu_tag[2] |= BIT(4); /* Enable VLAN forwarding offload */
 	h->cpu_tag[2] |= (vlan >> 8) & 0xf;
@@ -862,7 +862,7 @@ static void rtl839x_setup_notify_ring_buffer(struct rtl838x_eth_priv *priv)
 
 	/* Setup notification events */
 	sw_w32_mask(0, 1 << 14, RTL839X_L2_CTRL_0); /* RTL8390_L2_CTRL_0_FLUSH_NOTIFY_EN */
-	sw_w32_mask(0, 1 << 12, RTL839X_L2_NOTIFICATION_CTRL); /* SUSPEND_NOTIFICATION_EN
+	sw_w32_mask(0, 1 << 12, RTL839X_L2_NOTIFICATION_CTRL); /* SUSPEND_NOTIFICATION_EN */
 
 	/* Enable Notification */
 	sw_w32_mask(0, 1 << 0, RTL839X_L2_NOTIFICATION_CTRL);
@@ -1957,7 +1957,7 @@ static int rtl930x_mdio_reset(struct mii_bus *bus)
 			break;			/* Serdes: Value = 0 */
 		case PHY_INTERFACE_MODE_HSGMII:
 			private_poll_mask |= BIT(i);
-			/* fallthrough */
+			fallthrough;
 		case PHY_INTERFACE_MODE_USXGMII:
 			v |= BIT(mac_type_bit[i]);
 			uses_usxgmii = true;
diff --git a/target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c b/target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
index 77332527..cd513cc7 100644
--- a/target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
+++ b/target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
@@ -744,7 +744,7 @@ static int rtl8380_configure_int_rtl8218b(struct phy_device *phydev)
 	rtl8218b_6276B_hwEsd_perport = (void *)h + sizeof(struct fw_header) + h->parts[9].start;
 
 	if (sw_r32(RTL838X_DMY_REG31) == 0x1) {
-		int ipd_flag = 1;
+		int __attribute__((unused)) ipd_flag = 1;
 	}
 
 	val = phy_read(phydev, 0);
@@ -3330,7 +3330,7 @@ static void rtl931x_sds_rx_rst(u32 sds)
 	mdelay(50);
 }
 
-static void rtl931x_sds_disable(u32 sds)
+static __attribute__((unused)) void rtl931x_sds_disable(u32 sds)
 {
 	u32 v = 0x1f;
 
diff --git a/target/linux/realtek/patches-5.15/704-include-linux-add-phy-hsgmii-mode.patch b/target/linux/realtek/patches-5.15/704-include-linux-add-phy-hsgmii-mode.patch
index 8109abd2..213fe8a2 100644
--- a/target/linux/realtek/patches-5.15/704-include-linux-add-phy-hsgmii-mode.patch
+++ b/target/linux/realtek/patches-5.15/704-include-linux-add-phy-hsgmii-mode.patch
@@ -31,3 +31,18 @@ Submitted-by: Birger Koblitz <git@birger-koblitz.de>
  	case PHY_INTERFACE_MODE_QSGMII:
  		return "qsgmii";
  	case PHY_INTERFACE_MODE_TRGMII:
+diff --git a/drivers/net/phy/phylink.c b/drivers/net/phy/phylink.c
+index afe69295..4da98bc7 100644
+--- a/drivers/net/phy/phylink.c
++++ b/drivers/net/phy/phylink.c
+@@ -362,6 +362,10 @@ void phylink_get_linkmodes(unsigned long *linkmodes, phy_interface_t interface,
+ 		caps |= MAC_10000FD | MAC_5000FD | MAC_2500FD;
+ 		fallthrough;
+ 
++	case PHY_INTERFACE_MODE_HSGMII:
++		caps |= MAC_2500FD;
++		fallthrough;
++
+ 	case PHY_INTERFACE_MODE_RGMII_TXID:
+ 	case PHY_INTERFACE_MODE_RGMII_RXID:
+ 	case PHY_INTERFACE_MODE_RGMII_ID:
diff --git a/target/linux/realtek/patches-5.15/711-net-phy-add-an-MDIO-SMBus-library.patch b/target/linux/realtek/patches-5.15/711-net-phy-add-an-MDIO-SMBus-library.patch
index a80482d9..5c286b15 100644
--- a/target/linux/realtek/patches-5.15/711-net-phy-add-an-MDIO-SMBus-library.patch
+++ b/target/linux/realtek/patches-5.15/711-net-phy-add-an-MDIO-SMBus-library.patch
@@ -131,12 +131,12 @@ Signed-off-by: Antoine Tenart <antoine.tenart@bootlin.com>
 + * specified to be present in SFP modules.  These correspond with PHY
 + * addresses 16 and 17.  Disallow access to these "phy" addresses.
 + */
-+static bool i2c_mii_valid_phy_id(int phy_id)
++static __attribute__((unused)) bool i2c_mii_valid_phy_id(int phy_id)
 +{
 +	return phy_id != 0x10 && phy_id != 0x11;
 +}
 +
-+static unsigned int i2c_mii_phy_addr(int phy_id)
++static __attribute__((unused)) unsigned int i2c_mii_phy_addr(int phy_id)
 +{
 +	return phy_id + 0x40;
 +}
```